linear-operators/introduction-to-linear-operators/introduction-to-linear-operators.jmd

```diff
@@ -0,0 +1,301 @@
+---
+title: Introduction to LinearOperations.jl
+author: Abel Soares Siqueira
+---
+
+![](../../jso.png)
+
+```julia
+using Pkg
+pkg"activate ."
+if false
+  pkg"add LinearOperators"
+  pkg"add FFTW"
+  pkg"add Plots"
+  pkg"add PyPlot"
+end
+pkg"status"
+```
+
+LinearOperators.jl is a package for matrix-like operators. Linear operators are defined by how they act on a vector, which is useful in a variety of situations where you don't want to produce the matrix.
+
+```julia
+using LinearOperators
+
+prod(v) = [3v[1] - v[2]; 2v[1] + 2v[2]]
+#                  nrows, ncols, symm?, herm?, prod[; tprod, ctprod]
+A = LinearOperator(    2,     2, false, false, prod)
+
+A * ones(2)
+```
+
+The operator can be symmetric and/or Hermitian, and it the transpose and adjoint products might be defined as well:
+
+```julia
+prod(v) = [3v[1] - v[2]; 2v[1] + 2v[2]]
+tprod(v) = [3v[1] + 2v[2]; -v[1] + 2v[2]]
+
+A = LinearOperator(2, 2, false, false, prod, tprod)
+
+A' * ones(2)
+```
+
+```julia
+using FFTW, LinearAlgebra
+
+A = LinearOperator(16, 16, false, false, fft, nothing, ifft)
+
+v = rand(16) + im * rand(16)
+norm(A' * A * v - v)
+```
+
+Notice that in some cases missing operators can be inferred.
+
+```julia
+A = LinearOperator(16, 16, false, false, fft, nothing, ifft)
+v = rand(16) + im * rand(16)
+
+transpose(A) * v == conj.(ifft(conj.(v)))
+```
+
+## Lazy Products
+
+One immediate advantage of LinearOperators is that it allows lazy matrix products.
+
+```julia
+n = 4000
+A = rand(n, n)
+B = rand(n, n)
+opA = LinearOperator(A)
+opB = LinearOperator(B)
+
+# Run twice
+@time A * B
+@time opA * opB
+
+v = rand(n)
+@time A * (B * v)
+@time (opA * opB) * v;
+```
+
+## Inverse Operator
+
+Operators may be defined to represent (approximate) inverses.
+
+```julia
+A = rand(5,5)
+A = A' * A
+op = opCholesky(A)  # Use, e.g., as a preconditioner
+v = rand(5)
+
+norm(A \ v - op * v) / norm(v)
+```
+
+## LBFGS Operator
+
+One useful operator is the Limited BFGS operator, which implements the BFGS update with limited memory used in nonlinear optimization. This update comes in both direct ($B_k$) and inverse ($H_k = B_k^{-1}$) formats.
+$$
+B_{k+1} = B_k + \frac{y_k y_k^T}{y_k^T s_k} - \frac{B_k s_k s_k^T B_k^T}{s_k^T B_k s_k}
+$$
+$$
+H_{k+1} = \bigg(I - \frac{s_k y_k^T}{y_k^T s_k}\bigg)H_k\bigg(I - \frac{y_k s_k^T}{y_k^T s_k}\bigg) + \frac{s_k s_k^T}{y_k^T s_k}
+$$
```

Comment:
The math doesn't render on htmlpreview.github.io. Is there a MathJax js to be included somewhere?